src/Workspaces/CSharp/Portable/Extensions/SemanticModelExtensions.cs

```diff
@@ -451,9 +451,47 @@ public static ISet<INamespaceSymbol> GetUsingNamespacesInScope(this SemanticMode
                 }
             }
 
+            // For type constraints use declared accessibility of its member declaration (class, interface, delegate, method)
+            if (type.IsParentKind(SyntaxKind.TypeConstraint) &&
+                type.Parent.IsParentKind(SyntaxKind.TypeParameterConstraintClause))
+            {
+                var memberDeclaration = type.GetAncestor<MemberDeclarationSyntax>();
+                var declaredAccessibility = semanticModel.GetDeclaredSymbol(memberDeclaration, cancellationToken).DeclaredAccessibility;
+
+                return declaredAccessibility;
+            }
+
             return Accessibility.Private;
         }
 
+        public static bool AllContainingTypesArePublic(
+            this SemanticModel semanticModel,
+            TypeSyntax type,
+            CancellationToken cancellationToken)
+        {
+            if (type == null)
+            {
+                return false;
+            }
+
+            type = GetOutermostType(type);
+
+            var typeDeclarations = type.GetAncestors<TypeDeclarationSyntax>();
+
+            foreach (var typeDeclaration in typeDeclarations)
+            {
+                var symbol = semanticModel.GetDeclaredSymbol(typeDeclaration, cancellationToken);
+
+                if (symbol.DeclaredAccessibility == Accessibility.Private ||
+                    symbol.DeclaredAccessibility == Accessibility.Internal)
```

Comment:
This check doesn't match your method name.  For example, you'll return 'true' if one of the containing types i protected.  

That may be what you want, but in tha t case the method name should be AllContainingTypesArePublicOrProtected.